PyMca5/PyMcaGui/io/TiledCatalogSelector.py

```diff
@@ -0,0 +1,145 @@
+import logging
+from collections import defaultdict
+from typing import Callable, List, Mapping, Optional, Sequence
+from urllib.parse import ParseResult, urlparse as _urlparse
+
+from PyQt5.QtCore import QEvent, QObject, pyqtSignal
+
+from tiled.client import from_uri
+from tiled.client.base import BaseClient
+
+
+_logger = logging.getLogger(__name__)
+
+
+class TiledCatalogSelectorSignals(QObject):
+    """Collection of signals for a TiledCatalogSelector model."""
+    client_connected = pyqtSignal(
+        str, # new URL
+        str, # new API URL
+        name="TiledCatalogSelector.client_connected",
+    )
+    
+    def __init__(self, parent: Optional[QObject] = None) -> None:
+        super().__init__(parent)
+
+
+class TiledCatalogSelector(object):
+    """View Model for selecting a Tiled CatalogOfBlueskyRuns."""
+    Signals = TiledCatalogSelectorSignals
+
+    def __init__(
+        self,
+        /,
+        url: str = None,
+        client: BaseClient = None,
+        validators: Mapping[str, List[Callable]] = None,
+        parent: Optional[QObject] = None,
+        *args,
+        **kwargs,
+    ):
+        _logger.debug("TiledCatalogSelector.__init__()...")
+
+        self.url = url
+        self.client = client
+        self.validators = defaultdict(list)
+        if validators:
+            self.validators.update(validators)
+
+        self.signals = self.Signals(parent)
+        self.client_connected = self.signals.client_connected
+
+    def on_url_focus_in_event(self, event: QEvent):
+        """Handle the event when the URL widget gains focus."""
+        _logger.debug("TiledCatalogSelector.on_url_focus_in_event()...")
+
+        self._url_buffer = None
+
+    def on_url_text_edited(self, new_text: str):
+        """Handle a notification that the URL is being edited."""
+        _logger.debug("TiledCatalogSelector.on_url_text_edited()...")
+
+        self._url_buffer = new_text
+
+    def on_url_editing_finished(self):
+        """Handle a notification that URL editing is complete."""
+        _logger.debug("TiledCatalogSelector.on_url_editing_finished()...")
+
+        try:
+            for validate in self.validators["url"]:
+                validate(self._url_buffer)
+        except ValueError as exception:
+            _logger.info(str(exception))
+            # TODO: Should we re-raise the exception or emit a signal to inform the viewer?
```

Comment:
From group discussion with @AbbyGi & @danielballan :
* log an error
* send a signal about validation failing

Reply:
Resolved in commit `178818a4`